File: answers/7.Distributions/cat_heart_analysis.jl
# Cat Heart Weight Analysis
# Practical example: Fitting gamma distribution to cat heart weight data
# Uses method of moments and maximum likelihood estimation

using Distributions
using Random
using Statistics
using Plots
using Printf
using StatsBase
using HypothesisTests

# Load/simulate cat data (since MASS cats is R-specific)
# Using characteristics of the actual MASS cats dataset
Random.seed!(42)
n_cats = 144
cats_Hwt = rand(Gamma(17.5, 0.607), n_cats)  # Approximates real data

println("=== Cat Heart Weight Analysis ===\n")
println("Sample size: ", length(cats_Hwt))
@printf("Mean heart weight: %.4f g\n", mean(cats_Hwt))
@printf("SD heart weight: %.4f g\n", std(cats_Hwt))
@printf("Min: %.4f g, Max: %.4f g\n", minimum(cats_Hwt), maximum(cats_Hwt))

# Method of Moments estimation
println("\n--- Method of Moments Estimation ---")
function gamma_est_MM(x)
    m = mean(x)
    v = var(x)
    return (shape=m^2/v, scale=v/m)
end

mm_estimates = gamma_est_MM(cats_Hwt)
@printf("Shape: %.4f\n", mm_estimates.shape)
@printf("Scale: %.4f\n", mm_estimates.scale)
@printf("Rate: %.4f\n", 1/mm_estimates.scale)

# Maximum Likelihood Estimation
println("\n--- Maximum Likelihood Estimation ---")
mle_fit = fit(Gamma, cats_Hwt)
@printf("Shape: %.4f\n", shape(mle_fit))
@printf("Scale: %.4f\n", scale(mle_fit))
@printf("Rate: %.4f\n", 1/scale(mle_fit))

# Compare with the lesson's result
println("\n--- Comparison with Lesson Results ---")
println("Results may differ due to simulated data")
println("Lesson results: shape = 20.2998, rate = 1.9096")

# Create visualizations
p1 = histogram(cats_Hwt, bins=15, xlabel="Heart Weight (g)", ylabel="Frequency",
               title="Histogram of Cat Heart Weights", legend=false, alpha=0.7)

# Overlay fitted density on histogram
x_range = range(0, maximum(cats_Hwt)+2, length=200)
p2 = histogram(cats_Hwt, bins=15, normalize=:pdf, xlabel="Heart Weight (g)",
               ylabel="Density", title="Fitted Gamma Distribution",
               label="Data", alpha=0.7)
plot!(p2, x_range, pdf.(Gamma(mm_estimates.shape, mm_estimates.scale), x_range),
      linewidth=2, label="MM fit", color=:red)
plot!(p2, x_range, pdf.(mle_fit, x_range),
      linewidth=2, label="MLE fit", color=:blue)

# QQ plot
sorted_data = sort(cats_Hwt)
theoretical_quantiles = quantile.(Gamma(mm_estimates.shape, mm_estimates.scale),
                                 range(0.01, 0.99, length=length(cats_Hwt)))
p3 = scatter(theoretical_quantiles, sorted_data, xlabel="Theoretical Quantiles",
             ylabel="Observed Quantiles", title="QQ Plot (Method of Moments)",
             legend=false, alpha=0.6)
plot!(p3, [minimum(theoretical_quantiles), maximum(theoretical_quantiles)],
      [minimum(theoretical_quantiles), maximum(theoretical_quantiles)],
      linewidth=2, linestyle=:dash, color=:red)

# Kernel density estimation
kde_result = kde(cats_Hwt)
p4 = plot(kde_result.x, kde_result.density, linewidth=2, label="Kernel Density",
          xlabel="Heart Weight (g)", ylabel="Density",
          title="Kernel Density vs. Fitted Distribution")
plot!(p4, x_range, pdf.(Gamma(mm_estimates.shape, mm_estimates.scale), x_range),
      linewidth=2, linestyle=:dash, label="Fitted Gamma", color=:red)

plot(p1, p2, p3, p4, layout=(2,2), size=(1200, 1000))
savefig("/home/titan/pdfs/notes/statisticalComputingAndReporting/groupWork/answers/7.Distributions/cat_heart_analysis_jl.png")
println("\nVisualization saved as cat_heart_analysis_jl.png")

# Goodness of fit test (Kolmogorov-Smirnov)
println("\n--- Kolmogorov-Smirnov Test ---")
gamma_dist = Gamma(mm_estimates.shape, mm_estimates.scale)
ks_result = ExactOneSampleKSTest(cats_Hwt, gamma_dist)
@printf("KS statistic: %.6f\n", ks_result.δ)
@printf("p-value: %.4f\n", pvalue(ks_result))
if pvalue(ks_result) > 0.05
    println("Conclusion: Cannot reject H0. Data is consistent with Gamma distribution.")
else
    println("Conclusion: Reject H0. Data may not follow Gamma distribution.")
end

File: answers/7.Distributions/cat_heart_analysis.py
"""
Cat Heart Weight Analysis
Practical example: Fitting gamma distribution to cat heart weight data
Uses method of moments and maximum likelihood estimation
"""
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats

# Load cat data (simulated since MASS cats data is R-specific)
# The actual MASS cats dataset has 144 observations of cat heart weights
# We'll create similar data based on the reported statistics
np.random.seed(42)

# Based on MASS cats data characteristics
# Mean heart weight ≈ 10.63, SD ≈ 2.54
n_cats = 144
cats_Hwt = np.random.gamma(shape=17.5, scale=0.607, size=n_cats)  # Approximates real data

print("=== Cat Heart Weight Analysis ===\n")
print(f"Sample size: {len(cats_Hwt)}")
print(f"Mean heart weight: {np.mean(cats_Hwt):.4f} g")
print(f"SD heart weight: {np.std(cats_Hwt, ddof=1):.4f} g")
print(f"Min: {np.min(cats_Hwt):.4f} g, Max: {np.max(cats_Hwt):.4f} g")

# Visualize data
fig, axes = plt.subplots(2, 2, figsize=(12, 10))

# Histogram
axes[0, 0].hist(cats_Hwt, bins=15, edgecolor='black', alpha=0.7)
axes[0, 0].set_xlabel('Heart Weight (g)')
axes[0, 0].set_ylabel('Frequency')
axes[0, 0].set_title('Histogram of Cat Heart Weights')
axes[0, 0].grid(True, alpha=0.3)

# Method of Moments estimation
print("\n--- Method of Moments Estimation ---")
m = np.mean(cats_Hwt)
v = np.var(cats_Hwt, ddof=1)
mm_shape = m**2 / v
mm_scale = v / m

print(f"Shape: {mm_shape:.4f}")
print(f"Scale: {mm_scale:.4f}")
print(f"Rate: {1/mm_scale:.4f}")

# Maximum Likelihood Estimation
print("\n--- Maximum Likelihood Estimation ---")
shape_mle, loc_mle, scale_mle = stats.gamma.fit(cats_Hwt, floc=0)

print(f"Shape: {shape_mle:.4f}")
print(f"Scale: {scale_mle:.4f}")
print(f"Rate: {1/scale_mle:.4f}")

# Compare with the lesson's result
# From lesson: shape = 20.2998092, rate = 1.9095724
print("\n--- Comparison with Lesson Results ---")
print("Results may differ due to simulated data")
print("Lesson results: shape = 20.2998, rate = 1.9096")

# Overlay fitted density on histogram
x_range = np.linspace(0, np.max(cats_Hwt)+2, 200)
axes[0, 1].hist(cats_Hwt, bins=15, density=True, edgecolor='black', alpha=0.7, label='Data')
axes[0, 1].plot(x_range, stats.gamma.pdf(x_range, mm_shape, scale=mm_scale),
                'r-', linewidth=2, label='MM fit')
axes[0, 1].plot(x_range, stats.gamma.pdf(x_range, shape_mle, scale=scale_mle),
                'b-', linewidth=2, label='MLE fit')
axes[0, 1].set_xlabel('Heart Weight (g)')
axes[0, 1].set_ylabel('Density')
axes[0, 1].set_title('Fitted Gamma Distribution')
axes[0, 1].legend()
axes[0, 1].grid(True, alpha=0.3)

# QQ plot
theoretical_quantiles = np.linspace(0.01, 0.99, len(cats_Hwt))
theoretical_values = stats.gamma.ppf(theoretical_quantiles, mm_shape, scale=mm_scale)
observed_values = np.sort(cats_Hwt)

axes[1, 0].scatter(theoretical_values, observed_values, alpha=0.6)
axes[1, 0].plot([theoretical_values.min(), theoretical_values.max()],
                [theoretical_values.min(), theoretical_values.max()],
                'r--', linewidth=2)
axes[1, 0].set_xlabel('Theoretical Quantiles')
axes[1, 0].set_ylabel('Observed Quantiles')
axes[1, 0].set_title('QQ Plot (Method of Moments)')
axes[1, 0].grid(True, alpha=0.3)

# Kernel density estimation
from scipy.stats import gaussian_kde
kde = gaussian_kde(cats_Hwt)
x_kde = np.linspace(0, np.max(cats_Hwt)+2, 200)
axes[1, 1].plot(x_kde, kde(x_kde), 'k-', linewidth=2, label='Kernel Density')
axes[1, 1].plot(x_range, stats.gamma.pdf(x_range, mm_shape, scale=mm_scale),
                'r--', linewidth=2, label='Fitted Gamma')
axes[1, 1].set_xlabel('Heart Weight (g)')
axes[1, 1].set_ylabel('Density')
axes[1, 1].set_title('Kernel Density vs. Fitted Distribution')
axes[1, 1].legend()
axes[1, 1].grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig('/home/titan/pdfs/notes/statisticalComputingAndReporting/groupWork/answers/7.Distributions/cat_heart_analysis.png', dpi=150)
print("\nVisualization saved as cat_heart_analysis.png")
plt.close()

# Goodness of fit test (Kolmogorov-Smirnov)
print("\n--- Kolmogorov-Smirnov Test ---")
ks_stat, ks_pval = stats.kstest(cats_Hwt,
                                  lambda x: stats.gamma.cdf(x, mm_shape, scale=mm_scale))
print(f"KS statistic: {ks_stat:.6f}")
print(f"p-value: {ks_pval:.4f}")
if ks_pval > 0.05:
    print("Conclusion: Cannot reject H0. Data is consistent with Gamma distribution.")
else:
    print("Conclusion: Reject H0. Data may not follow Gamma distribution.")
